---
title: "Analysis of taxonomic abundance and diversity in lung microbiome"
author: "MDP"
date: "2023-08-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis of taxonomic abundance and diversity in lung microbiome

## Objectives

-   Separate datasets into samples with *Pseduomonas* spp. predominance or not
-   Define if predominance of *Pseduomonas* spp. in the microbiome of patients with Cystic Fibrosis(CF) and Non-Cystic fibrosis bronchiectasis (NCFB) affects the measures of diversity

## Preparation

### Command line: create summary files of taxonomy

The structure of the project repository is described below. All paths are relative to the `/script` folder where this script is located. The /processed_data folder contains all required data for downstream analysis.

```{bash Project dir structure, eval = FALSE}
.Rproject
├── notebook
├── processed_data
│   └── accessions
├── results
│   ├── tax_cf
│   └── tax_ncfb
└── scripts
```

After obtaining taxonomy classification using **nf-core** `taxprofiler_1.0.1` and `kraken2/bracken`, I use `kraken-biom_v1.2.0` to produce **biom** files necessary for analysis using `Phyloseq` in `R`.

```{bash Prepare .biom file, eval = FALSE}
  # starting at the root directory of repo
cd processed_data

  # taxpasta container is part of taxprofiler pipeline
for disease in {cf,ncfb}
    do 
    singularity exec $NXF_SINGULARITY_CACHE/kraken-biom_1.2.0.sif kraken-biom \
        --fmt json \
        --min G \
        -o ${disease}_kraken.biom \
        /scratch/mdprieto/results/cf_seed/taxprof_${disease}/bracken/k2_reports/* 
    done
```

### Prepare working environment

In case it is necessary, install the required packages.

```{r Packages install, eval = FALSE}
# ------- bioconductor

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")

# ------- base R

packages <- c("ggplot2", "RColorBrewer", "patchwork",
              "tidyverse", "flextable")

  # check if already installed
not_installed <- packages[!(packages %in% installed.packages()[ , "Package"])]

  # install only ones that are needed
if(length(not_installed)) install.packages(not_installed)    
```

Load all necessary libraries for analysis:

```{r Load packages, message = FALSE}
packages <- c("ggplot2", "RColorBrewer", "patchwork",
              "tidyverse", "phyloseq")
lapply(packages, require, character.only = TRUE)
```

### Load phyloseq objects and standardize them

Load **.biom** objects from `Bracken` taxonomic classification. Creates two 
`phyloseq` objects, one for each disease.

```{r Import .biom files}
cf_ps <- import_biom("processed_data/cf_bracken.biom")
ncfb_ps <- import_biom("processed_data/ncfb_bracken.biom")
```

Now, we do a few steps to normalize the data by adding proper rank names, changing 
the sample names to match the ones used in taxonomic summary, normalizing 
the abundance according to sequencing depth, excluding taxa with low abundance 
(increases variance), and calculating the relative abundance of each taxa.

```{r}
# ------- adjust names to match tsv file

sample_names(cf_ps) <- sub("_report", "", sample_names(cf_ps))
sample_names(ncfb_ps) <- sub("_report", "", sample_names(ncfb_ps))

# ------- clean taxonomy table

  # clean taxa names in CF phyloseq object
tax_table(cf_ps) <- sub(pattern = "[a-z]__", replacement = "", 
                        x =tax_table(cf_ps) )
  # add names of taxonomic ranks 
colnames(tax_table(cf_ps)) <- c("Kingdom", "Phylum", "Clade",
                                "Order", "Family", "Genus", "Species")
  
  # repeat cleaning process for NCFB phyloseq object
tax_table(ncfb_ps) <- sub("[a-z]__", "", tax_table(ncfb_ps))
colnames(tax_table(ncfb_ps)) <- c("Kingdom", "Phylum", "Clade",
                                  "Order", "Family", "Genus", "Species")

# ------- adjust to sequencing depth

  # CF
total = median(sample_sums(cf_ps))
cf_ps = transform_sample_counts(cf_ps,
                                function(x, t=total) round(t * (x / sum(x))))
  # NCFB
total = median(sample_sums(cf_ps))
ncfb_ps = transform_sample_counts(ncfb_ps,
                                  function(x, t=total) round(t * (x / sum(x))))

# ------- remove low abundance taxa

  # eliminate taxa with poor representation
cf_ps = filter_taxa(cf_ps,
                    function(x) sum(x > 3) > (0.2*length(x)), TRUE)
ncfb_ps = filter_taxa(ncfb_ps,
                      function(x) sum(x > 3) > (0.2*length(x)), TRUE)
```

### Import taxonomic summary table

Also import taxonomic classification summary, as it will be used to classify the samples where *Pseudomonas* spp. is the most abundant bacteria.

```{r Import taxonomic summaries}
# ------- import taxonomic summary excluding unnecessary columns
cf_summary_tsv <- read_tsv("results/tax_cf/CF_bracken_k2_db.tsv",
                           col_select = c(-3, -4))
ncfb_summary_tsv <- read_tsv("results/tax_ncfb/NCFB_bracken_k2_db.tsv",
                             col_select = c(-3, -4))

# ------- clean column names to reflect simple sample_id
colnames(cf_summary_tsv) <- sub(pattern = "_SR.*", replacement = "",
                               x = colnames(cf_summary_tsv))
colnames(ncfb_summary_tsv) <- sub("_SR.*", "", colnames(ncfb_summary_tsv))
```

## Predominance of *Pseudomonas* spp.

To make it easier to analyze the data, I pivot the table to a **long format** where every line has only one value assigned to an identifier. Furthermore, I calculate the relative abundance of a **taxa** in a new column.

Finally, I filter those cases where *Pseudomonas* spp. is the most abundant taxa in the microbiome and it represents more than 30% of the relative abundance.

```{r Predominance of Pseu in CF}
# ------- pivot to longer
  # do not pivot taxa_id columns and keep sample_ids together
cf_predominance <- 
  cf_summary_tsv |>
  pivot_longer(cols = c(-1,-2),
               cols_vary = "slowest",
               names_to = "sample_id",
               values_to = "assigned_reads") 
  
# ------- define predominance of pseudomonas

cf_predominance  <- 
  cf_predominance |> 
  dplyr::group_by(sample_id) |> 
    # calculate relative abundance
  dplyr::mutate(relative_abundance = assigned_reads / sum (assigned_reads)) |> 
    # remove human assigned reads
  dplyr::filter(!grepl("omo sapiens", name)) |> 
    # select top bacterial abundance
  dplyr::slice_max(n = 1, order_by = relative_abundance) |>
    # add factor to denote samples with predominance of pseudomonas
  dplyr::mutate(predominance = case_when(name == "Pseudomonas aeruginosa" & relative_abundance >= 0.3 ~ 'pseudomonas',
                                         TRUE ~ 'other'))

  # remove summary_tsv to keep environment tidy
rm(cf_summary_tsv)
```

The same process is performed for samples from NCFB patients

```{r Predominance of Pseu in NCFB}
# ------- pivot to long

ncfb_predominance <- 
  ncfb_summary_tsv |> 
  pivot_longer(cols = c(-1,-2),
               cols_vary = "slowest",
               names_to = "sample_id",
               values_to = "assigned_reads") 
  
# ------- define pseudomonas predominance for NCFB

ncfb_predominance  <- 
  ncfb_predominance |> 
  dplyr::group_by(sample_id) |> 
  dplyr::mutate(relative_abundance = assigned_reads / sum (assigned_reads)) |> 
  dplyr::filter(!grepl("omo sapiens", name)) |> 
  dplyr::slice_max(n = 1, order_by = relative_abundance) |>
  dplyr::mutate(predominance = case_when(name == "Pseudomonas aeruginosa" & relative_abundance >= 0.3 ~ 'pseudomonas',
                                         TRUE ~ 'other'))

  # remove summary_tsv to keep environment tidy
rm(ncfb_summary_tsv)
```

## Add factor variable of pseudomonas predominance to phyloseq object and create 
phyloseq objects with relative abundance

```{r}
  # CF
sample_data(cf_ps)$predominance <- cf_predominance$predominance

  # NCFB
sample_data(ncfb_ps)$predominance <- ncfb_predominance$predominance

  # relative abundance
cf_rel_abundance <- transform_sample_counts(cf_ps, 
                                            function(OTU) OTU/sum(OTU) )
ncfb_rel_abundance <- transform_sample_counts(ncfb_ps, 
                                              function(OTU) OTU/sum(OTU))
```

## Calculate diversity measures according to predominance group

### CF

Produce a summary bar plot of all species according to predominance variable

```{r}
cf_rel_abundance %>% 
  plot_bar(x = "Sample",
           fill = "Phylum") +
  facet_grid (~predominance, scale = "free_x") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  labs(title = expression("NCFB with"~italic(Pseudomonas)~"spp. predominance"))
```

I now calculate alpha diversity measurement and from the resulting plot obtain
the summarized estimates too

```{r}
cf_alpha <- 
  plot_richness(cf_ps, x = "predominance", color = "predominance",
              measures = c("Chao1", "Shannon", "InvSimpson")) +
  scale_colour_manual(values = c("red3", "blue3")) +
  theme_light() +
  theme(axis.text.x = element_blank()) +
  ggsignif::geom_signif(comparisons = list(c("pseudomonas", "other")),
                        map_signif_level = TRUE)
  ; cf_alpha

cf_alpha$data |> 
  pairwise.wilcox.test(predominance ~ variable)

#install.packages("ggsignif")
library(ggsignif)

t.test()
```


```{r}
estimate_richness(cf_predominance_pseu,
                  measures=c("Observed", "InvSimpson", "Shannon", "Chao1")) |> 
  summarise(mean_Chao1 = round(mean(Chao1), 2),
            mean_simpson = round(mean(InvSimpson), 2),
            mean_Shannon = round(mean(Shannon), 2)) |> 
  flextable::flextable()



estimate_richness(cf_predominance_other,
                  measures=c("Observed", "InvSimpson", "Shannon", "Chao1")) |> 
  summarise(mean_Chao1 = round(mean(Chao1), 2),
            mean_simpson = round(mean(InvSimpson), 2),
            mean_Shannon = round(mean(Shannon), 2)) |> 
  flextable::flextable()
```


### NCFB

```{r}
ncfb_predominance_pseu %>% 
  tax_glom("family") %>%  
  plot_bar(fill = "order",
           title="NCFB with _Pseudomonas_ spp. predominance")

```

Estimate richness and alpha diversity

```{r}
  # NCFB
estimate_richness(ncfb_predominance_pseu, 
                  measures=c("Observed", "InvSimpson", "Shannon", "Chao1")) |> 
  summarise(mean_Chao1 = round(mean(Chao1), 2),
            mean_simpson = round(mean(InvSimpson), 2),
            mean_Shannon = round(mean(Shannon), 2)) |> 
  flextable::flextable()
estimate_richness(ncfb_predominance_other,
                  measures=c("Observed", "InvSimpson", "Shannon", "Chao1")) |> 
  summarise(mean_Chao1 = round(mean(Chao1), 2),
            mean_simpson = round(mean(InvSimpson), 2),
            mean_Shannon = round(mean(Shannon), 2)) |> 
  flextable::flextable()

```


```{r}
plot_richness(cf_ps, x = "predominance", color = "predominance",
              measures = c("Chao1", "Shannon"))

  # CF pseudomonas
plot_richness(cf_predominance_pseu, x = "samples", 
              color = "predominance", measures = c("Chao1", "Shannon")) +
scale_colour_manual(values = c("red")) +
theme_light() + 
theme(axis.text.x = element_blank(),
      legend.position = "none")

  # CF other
plot_richness(cf_predominance_other, color = factor("red"),
             measures = c("Chao1", "Shannon")) +
scale_colour_manual(values = c("blue")) +
theme_light() + 
theme(axis.text.x = element_blank(),
      legend.position = "none")

  # NCFB pseudomonas
plot_richness(ncfb_predominance_pseu, color = factor("red"),
             measures = c("Chao1", "Shannon")) +
scale_colour_manual(values = c("red")) +
theme_light() + 
theme(axis.text.x = element_blank(),
      legend.position = "none")

  # NCFB other
plot_richness(ncfb_predominance_other, color = factor("red"),
             measures = c("Chao1", "Shannon")) +
scale_colour_manual(values = c("blue")) +
theme_light() + 
theme(axis.text.x = element_blank(),
      legend.position = "none")

```

